src/workerd/util/sqlite-metadata.c++

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) 2024 Cloudflare, Inc.
+// Licensed under the Apache 2.0 license found in the LICENSE file or at:
+//     https://opensource.org/licenses/Apache-2.0
+
+#include "sqlite-metadata.h"
+
+namespace workerd {
+
+SqliteMetadata::SqliteMetadata(SqliteDatabase& db) {
+  auto q = db.run("SELECT name FROM sqlite_master WHERE type='table' AND name='_cf_METADATA'");
+  if (q.isDone()) {
+    // The _cf_METADATA table doesn't exist. Defer initialization.
+    state = Uninitialized{db};
+  } else {
+    // The metadata table was initialized in the past. We can go ahead and prepare our statements.
+    // (We don't call ensureInitialized() here because the `CREATE TABLE IF NOT EXISTS` query it
+    // executes would be redundant.)
+    state = Initialized(db);
+  }
+}
+
+kj::Maybe<kj::Date> SqliteMetadata::getAlarm() {
+  auto& stmts = KJ_UNWRAP_OR(state.tryGet<Initialized>(), return kj::none);
+
+  auto query = stmts.stmtGetAlarm.run();
+  if (query.isDone() || query.isNull(0)) {
+    return kj::none;
+  } else {
+    return kj::UNIX_EPOCH + query.getInt64(0) * kj::NANOSECONDS;
+  }
+}
+
+void SqliteMetadata::setAlarm(kj::Maybe<kj::Date> currentTime) {
+  KJ_IF_SOME(t, currentTime) {
+    ensureInitialized().stmtSetAlarm.run((t - kj::UNIX_EPOCH) / kj::NANOSECONDS);
+  } else {
+    // Our getter code also allows representing an empty alarm value as a
+    // missing row or table, but a null-value row seems efficient and simple.
+    ensureInitialized().stmtSetAlarm.run(nullptr);
+  }
+}
+
+SqliteMetadata::Initialized& SqliteMetadata::ensureInitialized() {
+  KJ_SWITCH_ONEOF(state) {
+    KJ_CASE_ONEOF(uninitialized, Uninitialized) {
+      auto& db = uninitialized.db;
+
+      db.run(R"(
+        CREATE TABLE IF NOT EXISTS _cf_METADATA (
```

Comment:
I just merged #2680 which implements `deleteAll()` such that it deletes and recreates the entire database. That means after a `deleteAll()`, the _cf_METADATA table will be gone and needs to be recreated on-demand. You'll have to update this code to account for that, I guess.

Which raises the question: What should happen to the alarm on deleteAll()? With the old storage backend, the alarm is not actually deleted. To emulate that behavior, I suppose we have to recreate the alarm immediately, if it exists. We probably do need to emulate the behavior as we want to be able to migrate people transparently someday.

Relatedly, though, if an alarm is set on an object that has no other data in storage (whether because deleteAll() was called, or maybe nothing else was ever stored in the first place), and that alarm eventually runs and doesn't store anything nor set a new alarm... we need to make sure the object can be cleaned up. That won't happen unless we actually do another `deleteAll()` (or `db.reset()`) at that point.

We might need to add some code which runs whenever an alarm is unset, which checks if the database is completely empty, and resets it if it is...

That can probably be a separate PR though.

Reply:
* Added code reestablish the alarm state after `deleteAll()`, in 615deee2 .
* Added TODO to address database cleanup after final alarm is removed, in 2d833464 .